docs/subnets/introduction-to-vm.md

```diff
@@ -0,0 +1,335 @@
+
+## Blocks and State
+
+Virtual Machines can be broken down into 2 components: blocks and state. The functionality provided by VMs is to:
+
+- Define the representation of a blockchain's state
+- Represent the operations in that state
+- Apply the operations in that state
+
+Each block in the blockchain contains a set of state transitions. Each block is applied in order from the blockchain's initial genesis block to its last accepted block to reach the latest state of the blockchain.
+## Blockchain
+
+A blockchain has 2 components: **Consensus Engine** and **Virtual Machine**. VMs mainly deal with the implementation related to the block's structure: building and parsing. The Consensus Engine helps in reaching consensus on the block issued by VM. Here is a brief overview:
+
+1. A validator node wants to update the blockchain's state
+2. It creates a block with state transition details in it
+3. It issues the block to the consensus engine
+4. Consensus engine gossips the block within the network to reach consensus on it
+5. Depending upon the consensus results, the engine can either accept or reject the block
+6. Every virtuous node on the network should have the same verdict for a particular block
+7. It depends on the VM implementation on what to do with the accepted or rejected block
+
+The consensus engine has been already provided by `AvalancheGo` on the Avalanche network. Developers can implement and customize their own VM.
+
+## How to Load a VM
+
+VMs are created as a module, whose binary is registered by a node running `AvalancheGo`, against the **vmID** (binary file name must be vmID). VMID is a user-defined string that is zero-extended to a 32-byte array and encoded in CB58.
+
+The Avalanche nodes that want to power their chains with a particular VM, must have the built binary put in the proper place. See [here](../nodes/maintain/avalanchego-config-flags.md#--build-dir-string) for more details. There could be multiple VM plugins in this directory.
+
+We can build multiple blockchains with the registered VM. A VM is registered, only when its built binary is present at the required location.
+
+A blockchain can run as a separate process from AvalancheGo and can communicate with `AvalancheGo` over gRPC. This is enabled by `rpcchainvm`, a special VM that uses [`go-plugin`](https://pkg.go.dev/github.com/hashicorp/go-plugin) and wraps another VM implementation. The C-Chain, for example, runs the [Coreth](https://github.com/ava-labs/coreth) VM in this fashion.
+
+### APIs for a VM
+
+We interact with a blockchain and underlying VM through **Static** and **Non-Static Handlers**. Handlers are used for making API calls to VM methods. Specifically, these handlers are implemented as **Services**.
```

Comment:
Can we call them handlers and static handlers instead of static and non-static to coincide more directly with the function names on the VM?

I'm not a fan of calling them non-static as it makes static the default even though we barely use the static APIs.

Reply:
We should explain what a handler is in one sentence or call them HTTP handlers to make it more clear. Handler is a very vague term that can mean a lot of things.

Reply:
Can we distinguish handlers as blockchain handlers and VM handlers instead? Non-static handlers are only accessible on a blockchain, and static handlers on VM.

Reply:
Very nice. This sounds perfect to me ✌️.

Reply:
And I agree with Aaron that we should just call them APIs.